### ISSUE TITLE
Capture and replay a full decode step on the atoma-runtime tensor path

### ISSUE THREAD

**Issue by jorgeantonio21** [atoma-network/atoma-infer#143]
Part of #138

## Question

Can CUDA-graph capture work over candle's allocator/stream behavior, or does rung 2 need the `atoma-runtime::Tensor` path early? Locked call (plan §2.7): a 2–3 day capture-over-candle prototype immediately after PR-2, candle-vllm as reference. Clean capture + replay of a FullDecodeOnly step → ship graphs on candle in rung 2; fights (allocator churn, stream-capture violations, graph invalidation) → switch to the tensor path with evidence in hand.

Artifact: prototype branch + findings note (capture success matrix, replay latency vs eager, invalidation triggers).

**Comment by jorgeantonio21** [comment]
Re-scoped by #168 (2026-08-05): the capture-over-candle vs tensor-path question is decided — the decode hot path commits to atoma-runtime tensors, so this ticket is now the **tensor-path graph spike**. Scope: representative 2–4-layer decode step, 8B-shaped dims, op-class coverage (cuBLASLt GEMM, naive RMSNorm/RoPE/SwiGLU, FA2 via the candle-free ffi.rs, cache-write kernel, device argmax), world-size-1 NCCL capture check; static buffers + copy-in + replay. Acceptance: bucket-ladder capture (bs 1/8/32/64) incl. paged-KV writes; bit-identical replay logits vs eager; stable buffer addresses; ~1000-replay soak. Failures feed the rung-2 spec rather than deciding a substrate. Still blocked by #142; runs on the same H100 stint as the rung0-08 verification, immediately after it.

**Comment by jorgeantonio21** [comment]
Research phase complete (2026-08-05): four code-level reports (vLLM @ 8543522, SGLang @ e14fa10, TRT-LLM @ 1dfb7b1, native: llama.cpp/mistral.rs/candle-vllm/FlashInfer/cudarc-0.19.8) synthesized in the local mirror at docs/plan/research/cuda-graphs.md. Spike spec frozen at docs/plan/specs/143-graph-spike.md. Essentials, so this ticket is self-contained:

**Idiom confirmed:** static buffers + copy-in + replay (mistral.rs ships it default-on; llama.cpp abandoned per-step patching; cudaGraphExecUpdate is nobody's per-step mechanism — it's re-instantiation/segment repair only).

**Harness rules the spike must obey (top predicted failure sources):** cudarc `disable_event_tracking()` process-wide (per-launch event bookkeeping invalidates capture); RELAXED mode on a dedicated side stream; all allocs incl. NCCL init before first capture; no CudaSlice alloc/drop across a capture boundary; D2D-only per-step copy-in (captured pageable H2D = use-after-free in Rust); GraphEntry owns graph+exec+all buffers, drop order outputs → exec → comm; instantiate flags 0; wrap cuGraphDebugDotPrint day 1.

**Matrix:** bs {1,8,32,64} × {full decode step incl. KV write, +ws1 NCCL all-reduce}, replays with varying token ids/seq lens/block tables. Acceptance: the four #168 criteria. FA2 FFI path pre-audited clean (no alloc/sync/memcpy in flash_api.cu / cache_manager.cu / launch template / ffi.rs at develop b488cfa).

**Stretch (in order):** 2-segment breakable dry run (eager FA2 into a bridge buffer between captures); real weight-loaded 8B step.

Runs on the rung0-08 H100 stint immediately after the ~2h verification. Branch: feat/ja/graph-spike.

**Comment by jorgeantonio21** [comment]
Spike measurement note (2026-08-10): the per-graph memory measurement (dedicated overhead outside the shared pool, recorded per bucket) doubles as the calibration constant for rung-2 ladder budgeting. The activation arena footprint is computable in closed form; the driver-side `cudaGraphExec` state is not — so the graph-memory estimate for any candidate ladder is `measured per-graph overhead × ladder size`, and capturing 1–2 representative buckets suffices. No dry-run capture pass needed.

**Comment by jorgeantonio21** [comment]
Re-scoping now that #168 decided the substrate: this ticket no longer compares capture-over-candle against a tensor path — it validates the decided design before the production implementation is written.

Scope: capture a full Llama-8B-shaped decode step (qkv/o/mlp GEMMs, RMSNorm, RoPE, SwiGLU, paged decode attention via the FA2 kernels, KV-cache write, device argmax, and optionally a ws=1 NCCL all-reduce inside the captured step) and replay it against varying inputs. The harness is the `graph-harness` crate in #184.

Acceptance criteria (from #168):

1. Capture succeeds for every cell of the matrix: batch sizes 1/8/32/64, with and without the in-graph all-reduce.
2. Bit-identity: replayed logits identical to an eager run of the same ops across >=32 steps of varying token ids, sequence lengths, and block tables.
3. Stable addresses: every buffer address the graph baked is asserted unchanged at every replay.
4. Soak: ~1000 consecutive replays per bucket with zero free-memory delta after warm-up and no capture invalidation.

Failures classify as findings (kernel capture-illegal / cudarc behavior under capture / hidden FFI sync / NCCL), not as a verdict on the design. Deliverables: the capture-success matrix, host time per step eager vs replay, capture time, and per-graph dedicated memory.

**Comment by jorgeantonio21** [comment]
GPU findings

Environment: 1x NVIDIA H100 PCIe, compute capability 9.0, 114 SMs, driver
610.57.04, CUDA 12.8, NCCL 2.31.2, Ubuntu 24.04.

Harness: graph-harness at 6eb85f2a48043896f1c50f67d6b1236448500c37 (#184),
merged with #180, #181 and #182 at their current heads.

    cargo build -p graph-harness --features cuda,nccl --release --locked
    target/release/graph-harness run --nccl --out <dir>

Shape: 4 layers, hidden 4096, 32 query / 8 kv heads, head_dim 128, ffn 14336,
vocab 128256; page_block 16, max_seqlen 2048, start_seqlen 128, seed
1592614637.

## Acceptance criteria

1. Capture succeeds for every cell: 8 of 8, no failures.
2. Bit-identity: 32 of 32 steps per cell, no divergence.
3. Stable addresses: no drift. Every baked input, output, workspace and staging
   address is re-derived from the graph and compared after each identity step;
   a mismatch aborts the cell.
4. Soak: 1000 replays per cell, free-memory delta 0 after warm-up, no
   invalidation.

## Capture matrix

    cell        capture   nodes  identity  soak   mem delta
    bs64        ok  1 ms     79     32     1000       0
    bs64+nccl   ok  1 ms     87     32     1000       0
    bs32        ok  0 ms     79     32     1000       0
    bs32+nccl   ok  0 ms     87     32     1000       0
    bs8         ok  0 ms     79     32     1000       0
    bs8+nccl    ok  0 ms     87     32     1000       0
    bs1         ok  0 ms     83     32     1000       0
    bs1+nccl    ok  0 ms     91     32     1000       0

## Host time per step, microseconds (median / p99)

    cell         eager enqueue    replay enqueue   eager step     replay step
    bs64         417.6 / 755.4      9.9 / 16.8    2578.7 / 2914.8  2474.9 / 2485.5
    bs64+nccl    271.9 / 326.5      7.0 / 9.6     2601.6 / 2706.1  2497.0 / 2583.0
    bs32         247.1 / 274.0      6.3 / 14.2    2339.8 / 2368.2  2249.8 / 2298.9
    bs32+nccl    264.9 / 298.1      6.8 / 11.4    2362.5 / 2376.4  2259.5 / 2270.0
    bs8          244.3 / 273.9      6.7 / 11.4    2141.1 / 2156.2  2050.8 / 2065.7
    bs8+nccl     264.1 / 320.3      6.3 / 10.4    2166.5 / 2185.2  2065.7 / 2074.9
    bs1          227.2 / 264.7      3.6 / 8.7     2089.1 / 2101.2  1997.6 / 2000.9
    bs1+nccl     244.0 / 292.5      3.7 / 6.5     2110.1 / 2119.3  2011.3 / 2018.4

Replay cuts host enqueue time by 25-40x. Step time is dominated by device work
at these shapes, so the end-to-end gain is small here and will matter where the
step is enqueue-bound.

## Per-graph memory: the ladder budget is affine, not a product

The measurement note of 2026-08-10 assumed the ladder estimate is
"measured per-graph overhead x ladder size" and that capturing 1-2
representative buckets suffices. The run does not support either half.

Across the 8-cell matrix, graph_dedicated_bytes reports 33554432 for one cell
and 0 for the other seven. That is the driver's 2 MiB allocation granularity,
not a per-graph figure: the first instantiation absorbs a one-time reservation
and later graphs carve from it without moving the free-memory counter. Sampling
1-2 buckets would have produced either 32 MiB or 0 as the constant.

Re-running over 96 distinct buckets with --steps 1 --soak 0 amortizes the
granularity:

    first graph:            34 MiB   (one-time)
    95 further graphs:      22 MiB   total
    marginal per graph:    ~237 KiB

Budget for a ladder of N graphs: 34 MiB + ~237 KiB * N, rounded up to 2 MiB.
The fixed term dominates until roughly 150 graphs. Allocation events were
spaced 2 to 14 graphs apart, so 237 KiB is a mean rather than a bound, and a
ladder sized near a memory limit should carry margin for the quantization.

## Failure classification

None of the four predicted classes occurred: no capture-illegal kernel, no
cudarc misbehaviour under capture, no hidden FFI synchronisation, no NCCL
capture restriction. The ws=1 all-reduce captured inside the step with no
special handling.

The single defect was in the build graph rather than the capture path: the
harness binary did not link, because fa2.rs declares its own extern block and
nothing else in the crate named atoma-kernels, so cargo dropped the dependency
together with the native-library directives its build script emits.

    rust-lld: error: undefined symbol: run_mha
    rust-lld: error: undefined symbol: flash_last_error
    rust-lld: error: undefined symbol: flash_cuda_error_string
    rust-lld: error: undefined symbol: reshape_and_cache_flash_cache

## Not covered

- Segmented capture (two segments with an eager FA2 bridge between them) was
  not attempted. It remains the one part of the design with no evidence.
- Criterion 3 is enforced across the 32 identity steps, not across the 1000
  soak replays, which sample free memory only.
- Single GPU, so NCCL capture is exercised only at world size 1.
- A real weight-loaded 8B step was not run; weights are synthetic and bounded
  to +/-0.05 to keep bf16 activations in range.
